comment/managers/flags.py

```diff
@@ -0,0 +1,105 @@
+        (1, _('flagged')),
+        (2, _('flag rejected by moderator')),
+        (3, _('creator notified')),
+        (4, _('content removed by creator')),
+        (5, _('content removed by owner'))
+    ])
+
+    # Make a named tuple
+    State = namedtuple('State', ['value', 'state'])
+
+    # Construct the list of named tuples(can't use list comprehension, scope issues)
+    states = []
+    for state in STATES:
+        states.append(State(*state))
+
+    def get_flag_object(self, comment):
+        try:
+            flag = comment.flag
+        except ObjectDoesNotExist:
+            flag = self.create(comment=comment)
+
+        return flag
+
+
+class FlagInstanceManager(models.Manager):
+
+    REASONS = getattr(settings, 'COMMENT_FLAG_REASONS', [
+        (1, _('Spam | Exists only to promote a service')),
+        (2, _('Abusive | Intended at promoting hatred')),
+    ])
+    # add something else to the list
+    REASONS.append((100, _('Something else')))
+
+    # Make a named tuple
+    Reason = namedtuple('Reason', ['value', 'reason'])
+
+    # Construct the list of named tuples
+    reasons = []
+    for reason in REASONS:
+        reasons.append(Reason(*reason))
+
+    reason_values = [reason.value for reason in reasons]
+
+    def _clean_reason(self, reason):
+        err = ValidationError(
+                _('%(reason)s is an invalid reason'),
+                params={'reason': reason},
+                code='invalid'
+                )
+        try:
+            reason = int(reason)
+            if reason in self.reason_values:
+                return reason
+            raise err
+
+        except (ValueError, TypeError):
+            raise err
+
+    def _clean_action(self, action):
+        if isinstance(action, str):
+            act = action.lower()
+            if act in ['create', 'delete']:
+                return act
+
+        raise ValidationError(
+                _('%(action)s is not a valid action'),
+                params={'action': action},
+                code='invalid'
+                )
+
+    def _clean(self, reason, info):
+        reason = self._clean_reason(reason)
+        if reason == self.reason_values[-1] and (not info):
+            raise ValidationError(
+                _('Please supply some information as the reason for flagging'),
+                params={'info': info},
+                code='required'
+                )
+
+    def set_flag(self, user, flag, **kwargs):
+        reason = kwargs.get('reason', None)
+        info = kwargs.get('info', None)
+        action = kwargs.get('action', None)
+        self._clean_action(action)
+        if action == 'delete':
+            instance = self.get(flag=flag, user=user)
+            instance.delete()
+            created = False
+        else:
+            self._clean(reason, info)
+            self.create(flag=flag, user=user, reason=reason, info=info)
+            created = True
+
+        return created
```

Comment:
hey @abhiabhi94,
I have found an issue with the current design for flagging. The issue is derived from `action` variable when the user intends to flag a comment. IMHO we don't need this variable explicitly at all because the action is already known.
* If the user has not flagged a comment => the action will be `create`
* else `delete`
we can just implement a simple check to find out the action type instaed of passing a variable within JS and html and this will solve the api issue as well.
```suggestion
        reason = kwargs.get('reason', None)
        info = kwargs.get('info', None)
        instance, created = self.get_or_create(user=user, flag=flag)
        if not created:
            instance.delete()
        else:
            self._clean(reason, info)
            instance.reason = reason
            instance.info = info
            instance.save()

        return created
```

Reply:
I had used this to stop the abuse of the API. Now that I think of it, I don't think we can stop abuse just by doing this. Your suggestion seems legit to me. I will make the changes and remove the unnecessary variable. I will make the changes and push in a while.

Reply:
After making the changes, I now remember why I added the action argument. So in this case, for handling the wrong values of reason or info, we would have to catch the exception and delete the instance. This didn't seem like a nice way to deal with such cases to me. If you still want the change, I will push it.

Reply:
> So in this case, for handling the wrong values of reason or info, we would have to catch the exception and delete the instance.

I am not sure that I understood what you mean exactly, but what I see is that the action has nothing to do with reason and info here. As I said in case you need this variable, just keep it within django code and don't pass it from the JS or the template.
```python
instance, created = self.get_or_create(user=user, flag=flag)
if created:
    action = 'create'
```
I really don't see anymore reason for using the action variable since `created` is available for us.

There is an issue with the api when a user try to flag a comment, I will fix this after removing the action variable.

Reply:
Okay. Suppose a user tinkers something with the frontend and pass a wrong reason to the API. So according to your suggestion, `get_or_create` will create an instance and the execution will be passed to the else branch. Now the clean method throws a Validation error for the wrong value of reason. If put a `try except` block to catch this error, I will have to delete the instance now that this instance is no longer useful.

I didn't think this was a great method to deal with wrong values.

Reply:
> There is an issue with the api when a user try to flag a comment, I will fix this after removing the action variable.

What is the issue?

Reply:
> Okay. Suppose a user tinkers something with the frontend and passes a wrong reason to the API. So according to your suggestion, `get_or_create` will create an instance and the execution will be passed to the else branch. 

Simply check a valid reason before calling `get_or_create`.

Reply:
> What is the issue?

* How the data (reason, action and info) are passed when using api? You are using `**request.POST.dict()` but actually these data are not available in this dict, therefore, when you try to post to `comments/<int:pk>/flag/` you will always get an error, `None is not a valid action`
* Tried to pass these data as qurey string it worked somehow, but when posting twice with the same action, an `IntegrityError` is raised, therefore, I requested to remove action variable. IMHO it's not a good design.

Reply:
> * How the data (reason, action and info) are passed when using api? You are using `**request.POST.dict()` but actually these data are not available in this dict, therefore, when you try to post to `comments/<int:pk>/flag/` you will always get an error, `None is not a valid action`

Well, I just checked your tests and the data are passed correctly. The issue is that I used the GUI api of the restframework. This can be ignored now.

Reply:
> > Okay. Suppose a user tinkers something with the frontend and passes a wrong reason to the API. So according to your suggestion, `get_or_create` will create an instance and the execution will be passed to the else branch.
> 
> Simply check a valid reason before calling `get_or_create`.

For delete, no reason will be passed. This will create new issues.

Reply:
> > What is the issue?
> 
>     * How the data (reason, action and info) are passed when using api? You are using `**request.POST.dict()` but actually these data are not available in this dict, therefore, when you try to post to `comments/<int:pk>/flag/` you will always get an error, `None is not a valid action`
> 
>     * Tried to pass these data as qurey string it worked somehow, but when posting twice with the same action, an `IntegrityError` is raised, therefore, I requested to remove action variable. IMHO it's not a good design.

Do you get why the variable was used here?  Also, the parameters need to be passed via a query string.

Reply:
Delete the instance if no reason passed:
```python
        reason = kwargs.get('reason', None)
        info = kwargs.get('info', None)
        instance, created = self.get_or_create(user=user, flag=flag)
        if not created or not reason:
            instance.delete()
        else:
            self._clean(reason, info)
            instance.reason = reason
            instance.info = info
            instance.save()
        return created
```

Reply:
How do you deal with a wrong value of reason? You need to delete it in this case too.

Reply:
Also, you don't raise an error in the case when no reason is passed to flag a comment. Since you don't have an action variable, the code doesn't really know when to raise an exception.

Reply:
Ok then:
```python
        reason = kwargs.get('reason', None)
        info = kwargs.get('info', None)
        if reason:
             self._clean(reason, info)  # an exception will be raised if the reason not valid
        # here means that we have a valid reason or no reason passed
        instance, created = self.get_or_create(user=user, flag=flag)
        if not created or not reason:
            instance.delete()
        else:
            instance.reason = reason
            instance.info = info
            instance.save()
        return created
```

Reply:
Even in this case, the clean function will not run when no reason is passed to flag a comment. You want it to run everytime when flagging a comment, and let clean handle the exception handling.

Reply:
Why do I need to run clean if reason not passed. It must be abvoius that when there is no reason passed, the action must be delete if the instance exists.

Reply:
We have 2 cases only:
* reason is passed, the action is create if the reason valid, and.
* reason is not passed, the action is delete if the instance exists.

Reply:
Do we assume that there are no people in this universe who would tinker with the API? If someone tinkers with the API from the browser, an exception should be raised in my opinion if not all parameters are satisfied.

Reply:
I am not sure how this is related to our subject? However, you may leave it this way and I will refactore it before merging.

Reply:
Okay, I will repeat the example I gave above. 

If someone tinkers with the API and sends a request to the server without a reason should we just delete the instance or raise an exception if the parameters are not according to our need?

In this case, we delete the instance, because we have no way to know if the request for creation or deletion.

Reply:
> Okay, I will repeat the example I gave above.
> 
> If someone tinkers with the API and sends a request to the server without a reason should we just delete the instance or raise an exception if the parameters are not according to our need?

We will delete the instance, same user cannot flag a comment twice with a reason so this is abvoius that a request without a reason for particular user means deleting the flag.

> In this case, we delete the instance, because we have no way to know if the request was for creation or deletion.

We already know the type of the request. A flagged comment cannot be FLAGGED again by the same user, therefore, any further request from the same user means delete the existing flag.